File: Backend/app/url_eligibility.py
from __future__ import annotations

from dataclasses import dataclass
from urllib.parse import urlparse


UNSUPPORTED_HOST_SUFFIXES = (
    "youtube.com",
    "google.com",
    "gmail.com",
    "facebook.com",
    "instagram.com",
    "x.com",
    "twitter.com",
    "tiktok.com",
    "linkedin.com",
    "reddit.com",
    "pinterest.com",
    "mail.yahoo.com",
    "docs.google.com",
    "drive.google.com",
)

UNSUPPORTED_EXACT_PATHS = {
    "/",
    "/feed",
    "/shorts",
    "/search",
    "/results",
    "/explore",
    "/home",
    "/channel",
    "/login",
    "/signin",
    "/signup",
    "/settings",
    "/profile",
    "/account",
}

UNSUPPORTED_PREFIXES = (
    "/feed/",
    "/search/",
    "/results/",
    "/explore/",
    "/home/",
    "/channel/",
    "/login/",
    "/signin/",
    "/signup/",
    "/settings/",
    "/profile/",
    "/account/",
)

UNSUPPORTED_SPECIAL_PREFIXES = (
    "/watch",
    "/@",
)


@dataclass(frozen=True)
class UrlEligibility:
    is_supported: bool
    normalized_url: str | None
    reason_code: str
    reason_message: str


def _normalize_path(path: str) -> str:
    normalized = (path or "").strip()
    if not normalized:
        return "/"
    if not normalized.startswith("/"):
        normalized = f"/{normalized}"
    if len(normalized) > 1 and normalized.endswith("/"):
        normalized = normalized[:-1]
    return normalized or "/"


def _is_unsupported_host(hostname: str) -> bool:
    host = (hostname or "").strip().lower()
    if not host:
        return False
    if host.startswith("www."):
        host = host[4:]

    return any(host == suffix or host.endswith(f".{suffix}") for suffix in UNSUPPORTED_HOST_SUFFIXES)


def _looks_like_article_path(path: str) -> bool:
    normalized = _normalize_path(path)
    if normalized == "/":
        return False
    if normalized in UNSUPPORTED_EXACT_PATHS:
        return False
    if normalized.startswith(UNSUPPORTED_SPECIAL_PREFIXES):
        return False
    if any(normalized.startswith(prefix) for prefix in UNSUPPORTED_PREFIXES):
        return False

    segments = [segment for segment in normalized.split("/") if segment]
    if len(segments) >= 2:
        return True

    segment = segments[0] if segments else ""
    if len(segment) >= 16 and ("-" in segment or "_" in segment):
        return True

    return False


def classify_url_eligibility(url: str | None) -> UrlEligibility:
    candidate = (url or "").strip()
    if not candidate:
        return UrlEligibility(
            is_supported=False,
            normalized_url=None,
            reason_code="MISSING_URL",
            reason_message="No page URL was provided.",
        )

    parsed = urlparse(candidate)
    if parsed.scheme.lower() not in {"http", "https"}:
        return UrlEligibility(
            is_supported=False,
            normalized_url=None,
            reason_code="UNSUPPORTED_SCHEME",
            reason_message="Only regular web page URLs can be analyzed automatically.",
        )

    hostname = (parsed.hostname or "").strip().lower()
    if not hostname:
        return UrlEligibility(
            is_supported=False,
            normalized_url=None,
            reason_code="INVALID_HOST",
            reason_message="The page URL is missing a valid hostname.",
        )

    normalized_url = parsed.geturl()
    path = _normalize_path(parsed.path)

    if _is_unsupported_host(hostname):
        return UrlEligibility(
            is_supported=False,
            normalized_url=normalized_url,
            reason_code="UNSUPPORTED_HOST",
            reason_message="This site is treated as a non-article platform, so its page URL is ignored.",
        )

    if not _looks_like_article_path(path):
        return UrlEligibility(
            is_supported=False,
            normalized_url=normalized_url,
            reason_code="UNSUPPORTED_PATH",
            reason_message="This page URL does not look like an article page, so it is ignored.",
        )

    return UrlEligibility(
        is_supported=True,
        normalized_url=normalized_url,
        reason_code="SUPPORTED",
        reason_message="This page URL looks eligible for article analysis.",
    )

File: extension/url-eligibility.mjs
const UNSUPPORTED_HOST_SUFFIXES = [
  "youtube.com",
  "google.com",
  "gmail.com",
  "facebook.com",
  "instagram.com",
  "x.com",
  "twitter.com",
  "tiktok.com",
  "linkedin.com",
  "reddit.com",
  "pinterest.com",
  "mail.yahoo.com",
  "docs.google.com",
  "drive.google.com",
];

const UNSUPPORTED_EXACT_PATHS = new Set([
  "/",
  "/feed",
  "/shorts",
  "/search",
  "/results",
  "/explore",
  "/home",
  "/channel",
  "/login",
  "/signin",
  "/signup",
  "/settings",
  "/profile",
  "/account",
]);

const UNSUPPORTED_PREFIXES = [
  "/feed/",
  "/search/",
  "/results/",
  "/explore/",
  "/home/",
  "/channel/",
  "/login/",
  "/signin/",
  "/signup/",
  "/settings/",
  "/profile/",
  "/account/",
];

const UNSUPPORTED_SPECIAL_PREFIXES = ["/watch", "/@"];

function normalizePath(pathname) {
  let normalized = String(pathname || "").trim();
  if (!normalized) {
    return "/";
  }

  if (!normalized.startsWith("/")) {
    normalized = `/${normalized}`;
  }

  if (normalized.length > 1 && normalized.endsWith("/")) {
    normalized = normalized.slice(0, -1);
  }

  return normalized || "/";
}

function isUnsupportedHost(hostname) {
  let host = String(hostname || "").trim().toLowerCase();
  if (!host) {
    return false;
  }

  if (host.startsWith("www.")) {
    host = host.slice(4);
  }

  return UNSUPPORTED_HOST_SUFFIXES.some(
    (suffix) => host === suffix || host.endsWith(`.${suffix}`)
  );
}

function looksLikeArticlePath(pathname) {
  const normalized = normalizePath(pathname);
  if (normalized === "/") {
    return false;
  }

  if (UNSUPPORTED_EXACT_PATHS.has(normalized)) {
    return false;
  }

  if (UNSUPPORTED_SPECIAL_PREFIXES.some((prefix) => normalized.startsWith(prefix))) {
    return false;
  }

  if (UNSUPPORTED_PREFIXES.some((prefix) => normalized.startsWith(prefix))) {
    return false;
  }

  const segments = normalized.split("/").filter(Boolean);
  if (segments.length >= 2) {
    return true;
  }

  const segment = segments[0] || "";
  return segment.length >= 16 && (segment.includes("-") || segment.includes("_"));
}

export function classifyUrlEligibility(url) {
  const candidate = String(url || "").trim();
  if (!candidate) {
    return {
      isSupported: false,
      normalizedUrl: "",
      reasonCode: "MISSING_URL",
      reasonMessage: "Open a regular web page to include its URL in the analysis.",
    };
  }

  let parsed;
  try {
    parsed = new URL(candidate);
  } catch {
    return {
      isSupported: false,
      normalizedUrl: "",
      reasonCode: "INVALID_URL",
      reasonMessage: "The current page URL is invalid, so it will be ignored.",
    };
  }

  if (!["http:", "https:"].includes(parsed.protocol)) {
    return {
      isSupported: false,
      normalizedUrl: "",
      reasonCode: "UNSUPPORTED_SCHEME",
      reasonMessage: "Only regular web pages can be analyzed automatically.",
    };
  }

  if (!parsed.hostname) {
    return {
      isSupported: false,
      normalizedUrl: "",
      reasonCode: "INVALID_HOST",
      reasonMessage: "The current page URL has no valid hostname.",
    };
  }

  const normalizedUrl = parsed.toString();
  if (isUnsupportedHost(parsed.hostname)) {
    return {
      isSupported: false,
      normalizedUrl,
      reasonCode: "UNSUPPORTED_HOST",
      reasonMessage:
        "This site is treated as a non-article platform, so its URL will be ignored.",
    };
  }

  if (!looksLikeArticlePath(parsed.pathname)) {
    return {
      isSupported: false,
      normalizedUrl,
      reasonCode: "UNSUPPORTED_PATH",
      reasonMessage:
        "This page does not look like an article page, so its URL will be ignored.",
    };
  }

  return {
    isSupported: true,
    normalizedUrl,
    reasonCode: "SUPPORTED",
    reasonMessage: "Current page URL will be included in the analysis.",
  };
}
